internal/pkg/service/buffer/worker/service/cleanup.go

```diff
@@ -0,0 +1,90 @@
+package service
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/log"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/worker/distribution"
+)
+
+type cleanup struct {
+	*Service
+	logger   log.Logger
+	assigner *distribution.Assigner
+
+	lock *sync.RWMutex
+}
+
+func (s *Service) cleanup(ctx context.Context, wg *sync.WaitGroup) <-chan error {
+	return s.dist.StartWork(ctx, wg, s.logger, func(ctx context.Context, assigner *distribution.Assigner) (initDone <-chan error) {
+		return startCleanup(ctx, wg, s, assigner)
+	})
+}
+
+func startCleanup(ctx context.Context, wg *sync.WaitGroup, s *Service, assigner *distribution.Assigner) <-chan error {
+	c := &cleanup{
+		Service:  s,
+		logger:   s.logger.AddPrefix("[cleanup]"),
+		assigner: assigner,
+		lock:     &sync.RWMutex{},
+	}
+
+	// Start ticker
+	startTime := c.clock.Now()
+	c.startTicker(ctx, wg)
+	c.logger.Infof(`initialized | %s`, c.clock.Since(startTime))
+	initDone := make(chan error, 1)
+	defer close(initDone)
+	return initDone
+}
+
+// startTicker to check conditions periodically.
+func (c *cleanup) startTicker(ctx context.Context, wg *sync.WaitGroup) {
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		ticker := c.clock.Ticker(c.config.cleanupInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				c.check(ctx)
+			}
+		}
+	}()
+}
+
+func (c *cleanup) check(ctx context.Context) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	now := c.clock.Now()
+	receivers, err := c.store.ListAllReceivers(ctx)
+	if err != nil {
+		c.logger.Error(err)
+		return
+	}
+
+	for _, receiver := range receivers {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		err := c.store.Cleanup(ctx, receiver, c.logger)
+		if err != nil {
+			c.logger.Error(err)
+		}
+
+		time.Sleep(100 * time.Millisecond)
+	}
+
```

Comment:
Cez toto cakanie `100ms` nie je mozne operaciu ukoncit.

```suggestion
	for _, receiver := range receivers {
		err := c.store.Cleanup(ctx, receiver, c.logger)
		if err != nil && !errors.Is(err, context.Canceled) {
			c.logger.Error(err)
		}

		select {
		case <-ctx.Done():
			return
                case <- c.clock.TimeAfter(100 * time.Millisecond):
                        // let's wait a while to not overload etcd database
		}
	}
```

Ak sa `ctx` ukonci, tak request do `etcd` sa tiez ukonci a aj cele `c.store.Cleanup`, takze je jedno, ze sa kontext kontroluje na konci

Reply:
Tomu nerozumím, proč ji není možný dokončit? (To čekání jsi tam chtěl, aby se databáze moc nezahřívala.)

Reply:
To cakanie je v pohode, mam ho tiez v tom pripade.

No ten sleep bude vzdy cakat `100ms`, nic sa s tym neda spravit.

Toto bude cakat `100ms` alebo na koniec `ctx`, co nastane skor.
```
		select {
		case <-ctx.Done():
			return
                case <- c.clock.TimeAfter(100 * time.Millisecond):
                        // let's wait a while to not overload etcd database
		}
```

Samozrejme pri `100ms` to nie je problem, pri vacsom cakani by bol.
Je to anti-pattern (v testoch je to OK).

-------------

Je to detail, len v nejakej inej situacii by to mohol by problem "preco sa to zaseklo, ked som to ukoncil".
Napr. v CLI by si stlacil `CTRL+C`, ale program by bezal, kym ten sleep neskonci.

Reply:
https://stackoverflow.com/questions/55135239/how-can-i-sleep-with-responsive-context-cancelation

Reply:
Aha, ok.